packages/plugin-deduplicate/sources/commands/deduplicate.ts

```diff
@@ -0,0 +1,78 @@
+import * as semver                                      from 'semver';
+
+
+// eslint-disable-next-line arca/no-default-export
+export default class DeduplicateCommand extends BaseCommand {
+  static usage = Command.Usage({
+    category: `Workspace-related commands`,
+    description: `Reduces dependencies with overlapping ranges to a minimal set of packages`,
+    details: `https://github.com/atlassian/yarn-deduplicate for yarn v2`,
+    examples: [],
+  });
+
+  @Command.Path(`deduplicate`)
+  async execute() {
+    const configuration = await Configuration.find(this.context.cwd, this.context.plugins);
+    const {project} = await Project.find(configuration, this.context.cwd);
+
+    const locatorsByIdent: Map<IdentHash, Set<LocatorHash>> = new Map();
+    for (const [descriptorHash, locatorHash] of project.storedResolutions.entries()) {
+      const value = locatorHash;
+      const descriptor = project.storedDescriptors.get(descriptorHash)!
+      const key = descriptor.identHash;
+
+      const descriptors = locatorsByIdent.get(key);
+      if (descriptors === undefined) {
+        locatorsByIdent.set(key, new Set([value]));
+      } else {
+        locatorsByIdent.set(key, descriptors.add(value));
+      }
+    }
+
+    for (const descriptorHash of project.storedResolutions.keys()) {
+      const descriptor = project.storedDescriptors.get(descriptorHash)!
+      const locatorHashes = locatorsByIdent.get(descriptor.identHash)!;
+
+      if (locatorHashes !== undefined && locatorHashes.size > 1) {
+        const candidates = Array.from(locatorHashes).filter(candidateHash => {
+          const pkg  = project.storedPackages.get(candidateHash)!
+
+          if (structUtils.isVirtualLocator(pkg))
+            return false
+
+          if (pkg.version === null)
+            return false
+
+          const semverMatch = descriptor.range.match(/^npm:(.*)$/);
+          if (semverMatch === null)
+            return false;
+
+          return semver.satisfies(pkg.version, semverMatch[1])
+        }).sort((a, b) => {
+          const pkgA  = project.storedPackages.get(a)!
+          const pkgB  = project.storedPackages.get(b)!
+          return semver.gt(pkgA.version!, pkgB.version!) ? -1: 1;
+        });
+
+        if (candidates.length > 1) {
+          const newLocatorHash = candidates[0];
+          const oldLocatorHash = project.storedResolutions.get(descriptorHash)!
+          const newPkg = project.storedPackages.get(newLocatorHash)!;
+          const oldPkg = project.storedPackages.get(oldLocatorHash)!;
+
+          if (structUtils.areLocatorsEqual(oldPkg, newPkg) === false) {
+            console.log(`${structUtils.stringifyDescriptor(descriptor)} can be deduplicated from ${oldPkg.name}@${oldPkg.version} to ${newPkg.name}@${newPkg.version}`)
+
+            project.storedResolutions.set(descriptorHash, newLocatorHash);
+          }
+        }
+      }
+    }
+
+    await project.persistLockfile();
```

Comment:
I think this should be 
```suggestion
    await project.persist();
```
to also update .pnp.js

Reply:
Actually, since the `.pnp.js` file is a byproduct of the install, you need to run a complete install in order to update it (using `await project.install()`). Calling `persist` will just apply the changes you make to the manifest files (and since you don't make any, it won't have any effect).

Reply:
Replaced the project.persistLockfile with project.install. Has the nice side-effect of triggering postInstall hooks which clean the offline cache for this repo.